test/e2e/storage/testsuites/subpath.go

```diff
@@ -362,6 +362,21 @@ func testSubPath(input *subPathTestInput) {
 		testReadFile(input.f, input.filePathInSubpath, input.pod, 0)
 	})
 
+	It("should verify container cannot write to readonly volumes", func() {
+		if input.roVol == nil {
+			framework.Skipf("Volume type %v doesn't support readOnly source", input.volType)
+		}
+
+		// Format the volume while it's writable
+		formatVolume(input.f, input.formatPod)
+
+ 		// Set volume source to read only
+		input.pod.Spec.Volumes[0].VolumeSource = *input.roVol
+
+		// Write the file in the volume from container 1
+		setWriteCommand(input.subPathDir, &input.pod.Spec.Containers[1]) // HOW TO CHECK IF WRITE IS SUCCESSFUL? TODO
```

Comment:
I think this is fine.  The write should fail, causing the container to crash loop, which will cause the Pod to not become running.  testPodFailSubpathError() may need to be enhanced to not always check for the subpath error event.  In the case of CSI volumes, there would be no subpath error, but the container would be crash looping.

Reply:
For this particular subpath case, we want to test that the subpath volume is RO, so I would try to write to the container from container 0, instead of 1.

We can add a test case elsewhere to test readOnly for non-subpath volume mounts.

Reply:
@msau42 why should we write to container from container 0 instead of 1? Could you help me understand the difference?

Reply:
This test case is checking that we can't write to a subpath volume when the volume is readonly.  In the test pod, container 0 is the one that's configured to [use subpath](https://github.com/kubernetes/kubernetes/blob/a60ee42bb8c78c4a939b1f12163614723aaf301a/test/e2e/storage/testsuites/subpath.go#L452)

Reply:
@msau42 ah I see. For `testPodFailSubpathError()`, why would there be no subpath error for writing to subpath RO volumes? And what's the condition to/not to check for subpath error event? Thanks!!

Reply:
For CSI volumes, the underlying volume on the host [will not be mounted read only](https://github.com/kubernetes/kubernetes/pull/70662#issue-228457179), however the bind mount into the container will be read only.  This is different than how in-tree plugins behave, where the underlying host volume will be read only, in addition to the container bind mount.  But in the end, the user behavior is the same; they won't be able to write to the volume. 

So this test case will fail either by a subpath mounting error or by the container returning an error.

Reply:
@msau42 I see. So we shouldn't use `testPodFailSubpathError()`. I'm looking into how to assert that `f.TestContainerOutput(...)` (in `testReadFile`) fails but am not sure how to do that. I see that `f.TestContainerOutput(...)` calls `testContainerOutputMatcher` which has a `ExpectNoError` assertion which is essentially `ExpectWithOffset` and it returns an assertion. However, `f.TestContainerOutput(...)` doesn't return anything so I'm not sure how to assert the results of running `f.TestContainerOutput(...)`. Could you please give me some pointers? Thanks!!

Reply:
Sorry for the delay.  I recently reworked `testPodFailSubpathError` to check for pod conditions instead of events.  I think for this particular test case, we could add a new flag to [waitForPodSubpathError](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/test/e2e/storage/testsuites/subpath.go#L605:6) to also check for the container crashing in the container status.